### PR TITLE
[SPARK-44629][PYTHON][DOCS] Publish PySpark Test Guidelines webpage

### PR DESCRIPTION
**Pull request by asl3**
### What changes were proposed in this pull request?
This PR adds a webpage to the Spark docs website, https://spark.apache.org/docs, to outline PySpark testing best practices.

### Why are the changes needed?
The changes are needed to provide PySpark end users with a guideline for how to use PySpark utils (introduced in SPARK-44629) to test PySpark code.

### Does this PR introduce _any_ user-facing change?
Yes, the PR publishes a webpage on the Spark website.

### How was this patch tested?
Existing tests

### CODE REVIEW

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "9af4a35b-17e8-4e45-816b-34c14c5902f7",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Setting default log level to \"WARN\".\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:58]: I think you should remove this warning by setting `spark.sparkContext.setLogLevel("FATAL")` first, and remove that cell

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "a6c0b766-af5f-4e1d-acf8-887d7cf0b0b2",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:110]: This seems weird

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "8e533732-ee40-4cd0-9669-8eb92973908a",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/Users/amanda.liu/Documents/Databricks/spark/python/pyspark/pandas/__init__.py:50: UserWarning: 'PYARROW_IGNORE_TIMEZONE' environment variable was not set. It is required to set this environment variable to '1' in both driver and executor sides if you use pyarrow>=2.0.0. pandas-on-Spark will set it for you but it does not work if there is a Spark context already launched.\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:208]: hide this by the same. Set `PYARROW_IGNORE_TIMEZONE` in a cell first, and remove them to remove this warning

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ea1ee67d-db57-48c7-ba55-a32f64723e29",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# pytest /path/to/file"
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:414]: why comment?

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+    "        cls.spark = SparkSession.builder.appName(\"Sample PySpark ETL\").getOrCreate() \n",
+    "\n",
+    "    @classmethod\n",
+    "    def tearDownClass(cls):\n",
+    "        cls.spark.stop()\n",
+    "\n",
+    "\n",
+    "# Define unit test\n",
+    "class TestTranformation(PySparkTestCase):\n",
+    "    def test_single_space(self):\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:481]: indentation here looks weird

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+      "  warnings.warn(\n"
+     ]
+    }
+   ],
+   "source": [
+    "import pyspark.testing\n",
+    "\n",
+    "from pyspark.testing.utils import assertDataFrameEqual\n",
+    "\n",
+    "# Example 1\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:218]: separate cell?

File: python/docs/source/getting_started/index.rst [hunk 7]
```diff
@@ -40,3 +40,4 @@ The list below is the contents of this quickstart page:
    quickstart_df
    quickstart_connect
    quickstart_ps
+   testing_pyspark
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/index.rst:43]: Do you plan to create a separate binder link? or just keep this as a document without binder link?

Comment (asl3) [reply python/docs/source/getting_started/index.rst:43]: planning to keep as a document, with no binder link

**Comment by HyukjinKwon** [comment]
cc @xinrong-meng and @itholic for review.  see https://github.com/apache/spark/blob/1984c012f2f3dc2ec42b7a02183d2164a1c2922d/python/docs/source/getting_started/testing_pyspark.ipynb to review

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+   "source": [
+    "## Build a PySpark Application\n",
+    "Here is an example for how to start a PySpark application. Feel free to skip to the next section, “Testing your PySpark Application,” if you already have an application you’re ready to test.\n",
+    "\n",
+    "First, start your Spark Session."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:28]: can we start it from 1?

**Comment by xinrong-meng** [comment]
nit: Shall we consistently capitalize (or not) the first letter of each word in titles and subtitles for uniformity?

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d84a9fc1-9768-4af4-bfbf-e832f23334dc",
+   "metadata": {},
+   "source": [
+    "### Option 1: No test framework\n",
+    "\n",
+    "For simple ad-hoc validation cases, PySpark testing utils like assertDataFrameEqual and assertSchemaEqual can also be used in a standalone context.\n",
```
Comment (xinrong-meng) [inline python/docs/source/getting_started/testing_pyspark.ipynb:137]: nit: Shall we put `assertDataFrameEqual` and `assertSchemaEqual` in special fonts?

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c67be105-f6b1-4083-ad11-9e819331eae8",
+   "metadata": {},
+   "source": [
+    "### Option 2: Unit Test\n",
+    "For more complex testing scenarios, you may want to use a testing framework.\n",
+    "\n",
+    "One of the most popular testing framework options is unit tests. Let’s walk through how you can use the built-in Python unittest library to write PySpark tests. \n",
```
Comment (xinrong-meng) [inline python/docs/source/getting_started/testing_pyspark.ipynb:204]: Shall we attach the link of Python unittest official documentation?

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+    "\n",
+    "A better way is to write tests. Here are some examples of how we can test our code.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d84a9fc1-9768-4af4-bfbf-e832f23334dc",
+   "metadata": {},
+   "source": [
+    "### Option 1: No Test Framework\n",
```
Comment (allisonwang-db) [inline python/docs/source/getting_started/testing_pyspark.ipynb:135]: Option 1: No Test  -> Using the built-in PySpark Test Utility Functions

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+   "source": [
+    "When you run your test file with the `pytest` command, it will pick up all functions that have their name beginning with “test.”"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "940db965-3c2d-4b08-9c09-d2b7ee327c44",
+   "metadata": {},
+   "source": [
+    "## Option 4: Other Test Frameworks\n",
```
Comment (allisonwang-db) [inline python/docs/source/getting_started/testing_pyspark.ipynb:356]: We probably don't need this option :) We can simply mention at the beginning or the end of the section `## Testing your PySpark Application` that there are other ways to test PySpark applications besides the approaches mentioned in this example notebook.

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d8f50eee-5d0b-4719-b505-1b3ff05c16e8",
+   "metadata": {},
+   "source": [
+    "## Putting It All Together!\n",
+    "\n",
+    "Let’s see all the steps together, in a Unit Test example."
```
Comment (allisonwang-db) [inline python/docs/source/getting_started/testing_pyspark.ipynb:367]: Can we have this code in two separate Python cells (src/etl.py and tests/test_etl.py), one for the core logic and another for the unit test, and also include the command to run the unit test? It might be easier to use pytest instead of unittest.

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+    "transformed_df.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c471be1b-c052-4f31-abc9-35668aebc9c1",
+   "metadata": {},
+   "source": [
+    "You can also do this using Spark Connect. Please reference ‘Live Notebook: Spark Connect’ for examples.\n",
+    "For more information on how to use Spark Connect and its benefits, see: https://spark.apache.org/docs/latest/spark-connect-overview.html"
```
Comment (allisonwang-db) [inline python/docs/source/getting_started/testing_pyspark.ipynb:114]: Spark Connect here seems irrelevant.

**Review by allisonwang-db — COMMENTED** [review]
Looks good! Thanks for adding this!

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+   "execution_count": 21,
+   "id": "fa5db3a1-7305-44b7-ab84-f5ed55fd2ba9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pytest\n",
+    "from pyspark.testing.utils import assertDataFrameEqual\n",
+    "\n",
+    "def test_single_space(spark_fixture):\n",
+    "        sample_data = [{\"name\": \"John    D.\", \"age\": 30}, \n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:315]: indentation

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+    "spark = SparkSession.builder.appName(\"Sample PySpark ETL\").getOrCreate() \n",
+    "\n",
+    "sample_data = [{\"name\": \"John    D.\", \"age\": 30}, \n",
+    "  {\"name\": \"Alice   G.\", \"age\": 25}, \n",
+    "  {\"name\": \"Bob  T.\", \"age\": 35}, \n",
+    "  {\"name\": \"Eve   A.\", \"age\": 28}] \n",
+    "\n",
+    "df = spark.createDataFrame(sample_data)\n",
+    "\n",
+    "# define DataFrame transformation function\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:380]:
```suggestion
    "# Define DataFrame transformation function\n",
```

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+    "        \n",
+    "# Define unit test\n",
+    "class TestTranformation(PySparkTestCase):\n",
+    "    def test_single_space(self):\n",
+    "      sample_data = [{\"name\": \"John    D.\", \"age\": 30}, \n",
+    "    {\"name\": \"Alice   G.\", \"age\": 25}, \n",
+    "    {\"name\": \"Bob  T.\", \"age\": 35}, \n",
+    "    {\"name\": \"Eve   A.\", \"age\": 28}] \n",
+    "\n",
+    "    # create a Spark DataFrame\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:415]:
```suggestion
    "    # Create a Spark DataFrame\n",
```

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+    "    original_df = spark.createDataFrame(sample_data)\n",
+    "\n",
+    "    # apply the transformation function from before\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:418]:
```suggestion
    "    # Apply the transformation function from before\n",
```

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 19]
```diff
@@ -0,0 +1,486 @@
...
+    "# Define unit test base class\n",
+    "class PySparkTestCase(unittest.TestCase):\n",
+    "    @classmethod\n",
+    "    def setUpClass(cls):\n",
+    "        cls.spark = SparkSession.builder.appName(\"Sample PySpark ETL\").getOrCreate() \n",
+    "\n",
+    "    @classmethod\n",
+    "    def tearDownClass(cls):\n",
+    "        cls.spark.stop()\n",
+    "        \n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:406]: needs two newlines per PEP 8

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 20]
```diff
@@ -0,0 +1,486 @@
...
+   "outputs": [],
+   "source": [
+    "# src/etl.py\n",
+    "\n",
+    "from pyspark.sql import SparkSession \n",
+    "from pyspark.sql.functions import col\n",
+    "import pyspark.sql.functions as F \n",
+    "from pyspark.testing.utils import assertDataFrameEqual\n",
+    "\n",
+    "import unittest\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:368]: `import unittest` should be on the top per PEP 8

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 21]
```diff
@@ -0,0 +1,486 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "id": "fa5db3a1-7305-44b7-ab84-f5ed55fd2ba9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pytest\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:311]: one space after pytest per pep8

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 22]
```diff
@@ -0,0 +1,486 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "8e533732-ee40-4cd0-9669-8eb92973908a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pyspark.testing\n",
+    "\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:143]: remove newline

File: python/docs/source/getting_started/testing_pyspark.ipynb [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "id": "a2ea9dec-0ac0-4c23-8770-d6cc226d2e97",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# src/etl.py\n",
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/testing_pyspark.ipynb:361]: There are different practices how to place the source codes and test codes. One of the usual practice is to put `pkg/etl.py` and place tests together like `pkg/tests/test_etl.py`. I would fix the comments here.

**Comment by HyukjinKwon** [comment]
can we start this from [1] and remove the warnings?
Screenshot 2023-08-06 at 10 29 52 AM
Screenshot 2023-08-06 at 10 30 06 AM

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5.
